sui-execution/latest/sui-move-natives/src/transfer.rs

```diff
@@ -49,10 +52,22 @@ pub fn transfer_internal(
     let ty = ty_args.pop().unwrap();
     let recipient = pop_arg!(args, AccountAddress);
     let obj = args.pop_back().unwrap();
+    let object_is_shared = object_is_shared(context, &obj)?;
+
     let owner = Owner::AddressOwner(recipient.into());
-    object_runtime_transfer(context, owner, ty, obj)?;
+    let transfer_result = object_runtime_transfer(context, owner, ty, obj)?;
 
-    Ok(NativeResult::ok(context.gas_used(), smallvec![]))
+    let cost = context.gas_used();
+    if object_is_shared {
+        Ok(match transfer_result {
+            // New means the ID was created in this transaction
+            // SameOwner means the object was previously shared and was re-shared
+            TransferResult::New | TransferResult::SameOwner => NativeResult::ok(cost, smallvec![]),
```

Comment:
this will be the case whenever a shared object is created in the transaction, which is indeed ok

Reply:
Ah got it, I misunderstood what `TransferResult::New` meant (I thought it meant a new/different owner).  Thanks for clarifying.

Reply:
(...and now I see the comment that quite clearly explains what `New` means)

Reply:
To answer Todd's question, we've explicitly prevented adding shared objects as dynamic fields, so if we were to make such a sequence of events in one transaction, it would fail, as does the most recent test I've added

Reply:
@laura-makdah, preventing shared objects from being made into dynamic fields doesn't prevent the scenario that Todd described which is (in a single tx):

- Create an object,
- add it as a dynamic field of another object,
- remove it as a dynamic field,
- share it.

(i.e. share after dynamic field operations rather than the other way around).

Reply:
I've added the following test
```    
public fun make_dynamic_remove_and_then_share(ctx: &mut TxContext) {
        let child = Child { id: object::new(ctx), value: 0 };
        transfer::share_object(child);

        let parent = new_parent(ctx);

        add_field(&mut parent, child);
        let c: Child = ofield::remove(&mut parent.id, 0u64);
        transfer::share_object(c);
    }